<center>
    <h3>University of Toronto</h3>
    <h3>Department of Mechanical and Industrial Engineering</h3>
    <h3>MIE368 Analytics in Action </h3>
    <h3>(Fall 2020)</h3>
    <hr>
    <h1>Quiz One: Linear and Logistic Regression</h1>
    <h3>September 24, 2020</h3>
</center>


Question Bank 1 (2 marks)
---

>Q. Is **linear** regression appropriate for this application? Why?
>> Using linear regression is **not** appropriate. Linear regression predicts continuous target variables, but the target variables in this problem are binary.

> Q. Is **logistic** regression more appropriate for this application? Why?
>> Logistic regression is more appropriate. Logistic regression predicts binary target variables, and the target variables in this problem are binary.




Question 2. (2 marks)
---
> Q: What is overfitting? Are you worried about overfitting in this problem? Why?
>> Overfitting is when your model performs well on your training set and poorly on your test set. I am not worried about overfitting because the number of features (1) is much lower than the number of observations/samples (1667).


Question Bank 3 (1 mark)
---
> Q. Assume that the authors presented the confusion matrix below. What is the precision of their model? (2 possible matrices)
<figure>
<img src="https://docs.google.com/uc?export=view&id=1KJpXMy0dr4JVcGRIYdE6Dt5ypiXKnhHo"
alt="" />
<figcaption>Precision is 0.822</figcaption>
</figure>
<figure>

<img src="https://docs.google.com/uc?export=view&id=1AC38SWZxrHim7yoa-MZUsuMDGVS4moe4" 
alt=""/>
<figcaption>Precision is 0.741</figcaption>
</figure>


# Instructions

*   Please use this Colab notebook to solve the questions in the coding section of Quiz One. 
* Run the first codeblock to import the necessary packages and data for this quiz.
*  Any additional code you will need to answer the questions can be added to this notebook. Please remember to copy and paste the code at the bottom of the Quercus quiz and clearly label which question each piece of code belongs to. 
* Please note that in this quiz you will always test models on the same data that those models are trained on (i.e., use the full dataset to both train and test your model). This should help simplify your code :) 




In [2]:
# Run this code

# Packages used in this lab
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns
from sklearn.linear_model import LinearRegression, LogisticRegression, LassoCV, RidgeCV
from sklearn.metrics import confusion_matrix, roc_curve

# Import data for this quiz
df = pd.read_csv('https://docs.google.com/uc?export=download&id=1pzi2F_OLCUL96_wp5qGJ-JqM02bGf_eQ')
df = df.dropna(how='any')  # Drops the rows with missing data
df.head()  # Print a sample of the data

male  age  education  currentSmoker  ...    BMI  heartRate  glucose  TenYearCHD
0     1   39        4.0              0  ...  26.97       80.0     77.0           0
1     0   46        2.0              0  ...  28.73       95.0     76.0           0
2     1   48        1.0              1  ...  25.34       75.0     70.0           0
3     0   61        3.0              1  ...  28.58       65.0    103.0           1
4     0   46        3.0              1  ...  23.10       85.0     85.0           0

[5 rows x 16 columns]

In this quiz, you will use the Framingham Heart Study data set that you loaded in the previous code block. This data set contains the medical information of citizens from Framingham, MA. You will study characteristics that affect cardiovascular heart disease (CHD).

The columns in the dataset are as follows:


* **male**:            (binary) 1 if male, 0 if female
* **age**:             age in years    
* **education**:       (categorical) level of education (4 is highest)
* **currentSmoker**:   (binary) 1 if yes  
* **cigsPerDay**:      number of cigarettes smoked per day
* **BPMeds**:          (binary) 1 if they take blood pressure medication
* **prevalentStroke**: (binary) 1 if they have previously had a stroke
* **prevalentHyp**:    (binary) 1 if they prevalent hypertension
* **diabetes**:        (binary) 1 if they are diabetic
* **totChol**:         total cholesterol in mg/dL
* **sysBP (Target)**:           systolic blood pressure
* **diaBP**:           diastolic blood pressure in mmHg
* **BMI**:             Body Mass Index
* **heartRate**:       heart rate in beats/minute
* **glucose**:         glucose level in mg/dL  
* **TenYearCHD (Target)**:  (binary) risk of cardiovascular heart disease


Linear regression (5 marks)
---
In this question, you will use a simple linear regression model to predict the systolic blood pressure (`sysBP`) for a person. You should use `LinearRegression` with default parameters. Complete the following excercises:

>i) Train a model with the following features: ['male', 'age', 'diabetes', 'glucose', 'prevalentStroke', 'diaBP', 'currentSmoker']\\['male', 'age', 'diabetes', 'BPMeds', 'heartRate', 'diaBP', 'currentSmoker']\\['male', 'age', 'diabetes', 'totChol', 'prevalentHyp', 'diaBP', 'currentSmoker'].

>ii) What is the $R^2$ of your model?  (1 marks)

>iii) If someone [is a male (male)]\\[is on blood pressure medication (BPMeds)]\\[has previously had a stroke (prevalentStroke)][is currently a smoker (currentSmoker)]\\[is diabetic (diabetes)]\\[has prevalant hypertensions(prevalentHyp)], how much higher do you expect their systolic BP to be compared to someone who is not? (1 marks)  

>iv) What is the expected difference in systolic BP between a [40 year old male or a 30 year old female]\\[35 year old smoker or a 25 year old non-smoker]\\[a 80 year old diabetic or 50 year old non-diabetic]? Assume that they are otherwise identical. (1 marks)

In [3]:
# Write your code here.

# -------------------

# Part i
feature_list = ['male', 'age', 'diabetes', 'glucose', 'prevalentStroke', 'diaBP', 'currentSmoker']
# OR
# feature_list = ['male', 'age', 'diabetes', 'BPMeds', 'heartRate', 'diaBP', 'currentSmoker']
# feature_list = ['male', 'age', 'diabetes', 'totChol', 'prevalentHyp', 'diaBP', 'currentSmoker']
# Make dataset   
X = df[feature_list]
y = df.sysBP
# Train model
mdl = LinearRegression()
mdl.fit(X, y)

# Part ii
print(f'The R^2 is {mdl.score(X, y)}')

# Part iii
coefs = pd.DataFrame(mdl.coef_).T
coefs.columns=X.columns
male_difference = coefs['male'].abs()
# OR
smoker_difference = coefs['currentSmoker'].abs()
diabetic_difference = coefs['diabetes'].abs()

print(f'The difference is: \n \t {male_difference[0]} for males,\n\
\t {smoker_difference[0]} for current smokers,\n\
\t {diabetic_difference[0]} for diabetics')
# Part iv
male = coefs['male'].values + 40 * coefs['age'].values
female = 30 * coefs['age'].values
print(f'male has a BP that is {np.abs(male[0] - female[0])} higher than the female')

# OR
smoker = coefs['currentSmoker'].values + 35 * coefs['age'].values
non_smoker = 25 * coefs['age'].values
print(f'Smoker has a BP that is {np.abs(smoker[0] - non_smoker[0])} higher than the non-smoker')

diabetic = coefs['diabetes'].values + 80 * coefs['age'].values
non_diabetic = 50 * coefs['age'].values
print(f'Diabetic has a BP that is {np.abs(diabetic[0] - non_diabetic[0])} higher than the non-diabetic')

# -------------------

The R^2 is 0.6817304102128598
The difference is: 
 	 3.6747840137886603 for males,
	 0.897230615113782 for current smokers,
	 0.8843024417223241 for diabetics
male has a BP that is 2.2115145666749605 higher than the female
Smoker has a BP that is 6.7835291955773975 higher than the non-smoker
Diabetic has a BP that is 18.543198183113184 higher than the non-diabetic


In [4]:

# Part i
feature_list = ['male', 'age', 'diabetes', 'glucose', 'prevalentStroke', 'diaBP', 'currentSmoker']
# OR
# feature_list = ['male', 'age', 'diabetes', 'BPMeds', 'heartRate', 'diaBP', 'currentSmoker']
# feature_list = ['male', 'age', 'diabetes', 'totChol', 'prevalentHyp', 'diaBP', 'currentSmoker']
# Make dataset   
X = df.filter(feature_list)

___
## __Answer__:

**Part i)** 0.682\0.689\0.503

**Part ii)** -2.08\25.34\2.34\0.04\-0.08\32.15

**Part iii)** Male higher by 6.26\Smoker is higher by 6.08\Diabetic is higher by 5.39

___

## Logistic regression (5 marks)

In this question, you will use logistic regression to predict risk of heart disease (`TenYearCHD`) using all of the other data as features. Complete the following excercises:

1. Fit a logistic regression model is [a maximum iteration limit of 3250 and the sag solver]/[a l1 penalty and the liblinear solver]
2. What is the accuracy of your model?
3. What is the odds that a person who [previously had a stroke]/[is a male]/[is currently a smoker] also has a risk of heart disease?
4. What threshold should you choose for your model in order to achieve a [90% true positive rate/5% false positive rate]. 

In [5]:
# Write your code here.

# -------------------

X = df.drop(columns=['TenYearCHD'])
y = df.TenYearCHD

# Part 1
mdl = LogisticRegression(max_iter=3250,solver='sag')
# OR
# mdl = LogisticRegression(penalty='l1', solver='liblinear')
mdl.fit(X, y)

# Part 2 
print(f'The accuracy of the model is {mdl.score(X, y):.3f}')

# Part 3
coefs = pd.DataFrame(mdl.coef_, columns=X.columns)
person_who = 'prevalentStroke'  
# OR 
# person_who = 'male'  
# person_who = 'currentSmoker'  
odds = np.exp(coefs[person_who][0])
print(f'The odds for a person who is a {person_who} is {odds:.3f}')

# Part 4
y_pred_prob = mdl.predict_proba(X).T[1]  # Predictions without threshold 
fpr, tpr, threshold = roc_curve(y, y_pred_prob)  
tpr_target_high = np.argwhere(tpr>=0.9)[0,0]
tpr_target_low = np.argwhere(tpr>=0.9)[-1,0]
fpr_target = np.argwhere(fpr>=0.05)[0,0]
new_tpr_treshold_low = threshold[tpr_target_low]
new_tpr_treshold_high= threshold[tpr_target_high]
new_fpr_treshold = threshold[fpr_target]
print(f'The lowest threshold that achieves 90% TPR is {new_tpr_treshold_low:.5}')
print(f'The highest threshold that achieves 90% TPR is {new_tpr_treshold_high:.5}')
print(f'The lowest threshold that achieves 5% FPR is {new_fpr_treshold:.5}')
# -------------------

The accuracy of the model is 0.851
The odds for a person who is a prevalentStroke is 1.086
The lowest threshold that achieves 90% TPR is 0.012963
The highest threshold that achieves 90% TPR is 0.084471
The lowest threshold that achieves 5% FPR is 0.3287
